{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c479dc3e",
   "metadata": {},
   "source": [
    "# Python imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "358cb429",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show plots as part of the notebook (this is a Jupyter-specific operation)\n",
    "%matplotlib inline\n",
    "import time\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Standard library packages\n",
    "import os\n",
    "\n",
    "# Import Pandas and Seaborn\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Import Biopython tools for running local BLASTX\n",
    "from Bio.Blast.Applications import NcbiblastxCommandline\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5a4a36dc",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'blastn' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_949863/169391034.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mblastn\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0mhelp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'blastn' is not defined"
     ]
    }
   ],
   "source": [
    "blastn -help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "675915d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class NcbiblastnCommandline in module Bio.Blast.Applications:\n",
      "\n",
      "class NcbiblastnCommandline(_NcbiblastMain2SeqCommandline)\n",
      " |  NcbiblastnCommandline(cmd='blastn', **kwargs)\n",
      " |  \n",
      " |  Wrapper for the NCBI BLAST+ program blastn (for nucleotides).\n",
      " |  \n",
      " |  With the release of BLAST+ (BLAST rewritten in C++ instead of C), the NCBI\n",
      " |  replaced the old blastall tool with separate tools for each of the searches.\n",
      " |  This wrapper therefore replaces BlastallCommandline with option -p blastn.\n",
      " |  \n",
      " |  For example, to run a search against the \"nt\" nucleotide database using the\n",
      " |  FASTA nucleotide file \"m_code.fasta\" as the query, with an expectation value\n",
      " |  cut off of 0.001, saving the output to a file in XML format:\n",
      " |  \n",
      " |  >>> from Bio.Blast.Applications import NcbiblastnCommandline\n",
      " |  >>> cline = NcbiblastnCommandline(query=\"m_cold.fasta\", db=\"nt\", strand=\"plus\",\n",
      " |  ...                               evalue=0.001, out=\"m_cold.xml\", outfmt=5)\n",
      " |  >>> cline\n",
      " |  NcbiblastnCommandline(cmd='blastn', out='m_cold.xml', outfmt=5, query='m_cold.fasta', db='nt', evalue=0.001, strand='plus')\n",
      " |  >>> print(cline)\n",
      " |  blastn -out m_cold.xml -outfmt 5 -query m_cold.fasta -db nt -evalue 0.001 -strand plus\n",
      " |  \n",
      " |  You would typically run the command line with cline() or via the Python\n",
      " |  subprocess module, as described in the Biopython tutorial.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      NcbiblastnCommandline\n",
      " |      _NcbiblastMain2SeqCommandline\n",
      " |      _Ncbiblast2SeqCommandline\n",
      " |      _NcbiblastCommandline\n",
      " |      _NcbibaseblastCommandline\n",
      " |      Bio.Application.AbstractCommandline\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, cmd='blastn', **kwargs)\n",
      " |      Initialize the class.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  best_hit_overhang\n",
      " |      Best Hit algorithm overhang value (float, recommended value: 0.1)\n",
      " |      \n",
      " |      Float between 0.0 and 0.5 inclusive. Incompatible with: culling_limit.\n",
      " |      \n",
      " |      This controls the addition of the -best_hit_overhang parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  best_hit_score_edge\n",
      " |      Best Hit algorithm score edge value (float).\n",
      " |      \n",
      " |      Float between 0.0 and 0.5 inclusive. Recommended value: 0.1\n",
      " |      \n",
      " |      Incompatible with: culling_limit.\n",
      " |      \n",
      " |      This controls the addition of the -best_hit_score_edge parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  culling_limit\n",
      " |      Hit culling limit (integer).\n",
      " |      \n",
      " |      If the query range of a hit is enveloped by that of at least this many higher-scoring hits, delete the hit.\n",
      " |      \n",
      " |      Incompatible with: best_hit_overhang, best_hit_score_edge.\n",
      " |      \n",
      " |      This controls the addition of the -culling_limit parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  db\n",
      " |      The database to BLAST against.\n",
      " |      \n",
      " |      This controls the addition of the -db parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  db_hard_mask\n",
      " |      Filtering algorithm for hard masking (integer).\n",
      " |      \n",
      " |      Filtering algorithm ID to apply to BLAST database as hard masking. Incompatible with: db_soft_mask, subject, subject_loc\n",
      " |      \n",
      " |      This controls the addition of the -db_hard_mask parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  db_soft_mask\n",
      " |      Filtering algorithm for soft masking (integer).\n",
      " |      \n",
      " |      Filtering algorithm ID to apply to BLAST database as soft masking. Incompatible with: db_hard_mask, subject, subject_loc\n",
      " |      \n",
      " |      This controls the addition of the -db_soft_mask parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  dbsize\n",
      " |      Effective length of the database (integer).\n",
      " |      \n",
      " |      This controls the addition of the -dbsize parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  dust\n",
      " |      Filter query sequence with DUST (string).\n",
      " |      \n",
      " |      Format: 'yes', 'level window linker', or 'no' to disable.\n",
      " |      \n",
      " |      Default = '20 64 1'.\n",
      " |      \n",
      " |      This controls the addition of the -dust parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  entrez_query\n",
      " |      Restrict search with the given Entrez query (requires remote).\n",
      " |      \n",
      " |      This controls the addition of the -entrez_query parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  evalue\n",
      " |      Expectation value cutoff.\n",
      " |      \n",
      " |      This controls the addition of the -evalue parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  export_search_strategy\n",
      " |      File name to record the search strategy used.\n",
      " |      \n",
      " |      Incompatible with: import_search_strategy\n",
      " |      \n",
      " |      This controls the addition of the -export_search_strategy parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  filtering_db\n",
      " |      BLAST database containing filtering elements (i.e. repeats).\n",
      " |      \n",
      " |      This controls the addition of the -filtering_db parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  gapextend\n",
      " |      Cost to extend a gap (integer).\n",
      " |      \n",
      " |      This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  gapopen\n",
      " |      Cost to open a gap (integer).\n",
      " |      \n",
      " |      This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  gilist\n",
      " |      Restrict search of database to list of GI's.\n",
      " |      \n",
      " |      Incompatible with: negative_gilist, seqidlist, negative_seqidlist, remote, subject, subject_loc\n",
      " |      \n",
      " |      This controls the addition of the -gilist parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  h\n",
      " |      Print USAGE and DESCRIPTION;  ignore other arguments.\n",
      " |      \n",
      " |      This property controls the addition of the -h switch, treat this property as a boolean.\n",
      " |  \n",
      " |  help\n",
      " |      Print USAGE, DESCRIPTION and ARGUMENTS description; ignore other arguments.\n",
      " |      \n",
      " |      This property controls the addition of the -help switch, treat this property as a boolean.\n",
      " |  \n",
      " |  html\n",
      " |      Produce HTML output? See also the outfmt option.\n",
      " |      \n",
      " |      This property controls the addition of the -html switch, treat this property as a boolean.\n",
      " |  \n",
      " |  import_search_strategy\n",
      " |      Search strategy to use.\n",
      " |      \n",
      " |      Incompatible with: export_search_strategy\n",
      " |      \n",
      " |      This controls the addition of the -import_search_strategy parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  index_name\n",
      " |      MegaBLAST database index name.\n",
      " |      \n",
      " |      This controls the addition of the -index_name parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  lcase_masking\n",
      " |      Use lower case filtering in query and subject sequence(s)?\n",
      " |      \n",
      " |      This property controls the addition of the -lcase_masking switch, treat this property as a boolean.\n",
      " |  \n",
      " |  line_length\n",
      " |      Line length for formatting alignments (integer, at least 1, default 60).\n",
      " |      \n",
      " |      Not applicable for outfmt > 4. Added in BLAST+ 2.2.30.\n",
      " |      \n",
      " |      This controls the addition of the -line_length parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  max_hsps\n",
      " |      Set max number of HSPs saved per subject sequence\n",
      " |      \n",
      " |      Ddefault 0 means no limit.\n",
      " |      \n",
      " |      This controls the addition of the -max_hsps parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  max_hsps_per_subject\n",
      " |      Override max number of HSPs per subject saved for ungapped searches (integer).\n",
      " |      \n",
      " |      This controls the addition of the -max_hsps_per_subject parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  max_target_seqs\n",
      " |      Maximum number of aligned sequences to keep (integer, at least one).\n",
      " |      \n",
      " |      This controls the addition of the -max_target_seqs parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  min_raw_gapped_score\n",
      " |      Minimum raw gapped score to keep an alignment in the preliminary gapped and traceback stages (integer).\n",
      " |      \n",
      " |      This controls the addition of the -min_raw_gapped_score parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  negative_gilist\n",
      " |      Restrict search of database to everything except the listed GIs.\n",
      " |      \n",
      " |      Incompatible with: gilist, seqidlist, remote, subject, subject_loc\n",
      " |      \n",
      " |      This controls the addition of the -negative_gilist parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  negative_seqidlist\n",
      " |      Restrict search of database to everything except listed SeqID's.\n",
      " |      \n",
      " |      Incompatible with: gilist, seqidlist, remote, subject, subject_loc\n",
      " |      \n",
      " |      This controls the addition of the -negative_seqidlist parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  no_greedy\n",
      " |      Use non-greedy dynamic programming extension\n",
      " |      \n",
      " |      This property controls the addition of the -no_greedy switch, treat this property as a boolean.\n",
      " |  \n",
      " |  num_alignments\n",
      " |      Number of database sequences to show num_alignments for.\n",
      " |      \n",
      " |      Integer argument (at least zero). Default is 200. See also num_alignments.\n",
      " |      \n",
      " |      This controls the addition of the -num_alignments parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  num_descriptions\n",
      " |      Number of database sequences to show one-line descriptions for.\n",
      " |      \n",
      " |      Integer argument (at least zero). Default is 500. See also num_alignments.\n",
      " |      \n",
      " |      This controls the addition of the -num_descriptions parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  num_threads\n",
      " |      Number of threads to use in the BLAST search.\n",
      " |      \n",
      " |      Integer, at least one. Default is one. Incompatible with: remote\n",
      " |      \n",
      " |      This controls the addition of the -num_threads parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  off_diagonal_range\n",
      " |      Number of off-diagonals to search for the 2nd hit (integer).\n",
      " |      \n",
      " |      Expects a positive integer, or 0 (default) to turn off.Added in BLAST 2.2.23+\n",
      " |      \n",
      " |      This controls the addition of the -off_diagonal_range parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  out\n",
      " |      Output file for alignment.\n",
      " |      \n",
      " |      This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  outfmt\n",
      " |      Alignment view.  Typically an integer 0-14 but for some formats can be named columns like '6 qseqid sseqid'.  Use 5 for XML output (differs from classic BLAST which used 7 for XML).\n",
      " |      \n",
      " |      This controls the addition of the -outfmt parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  parse_deflines\n",
      " |      Should the query and subject defline(s) be parsed?\n",
      " |      \n",
      " |      This property controls the addition of the -parse_deflines switch, treat this property as a boolean.\n",
      " |  \n",
      " |  penalty\n",
      " |      Penalty for a nucleotide mismatch (integer, at most zero).\n",
      " |      \n",
      " |      This controls the addition of the -penalty parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  perc_identity\n",
      " |      Percent identity (real, 0 to 100 inclusive).\n",
      " |      \n",
      " |      This controls the addition of the -perc_identity parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  qcov_hsp_perc\n",
      " |      Percent query coverage per hsp (float, 0 to 100).\n",
      " |      \n",
      " |      Added in BLAST+ 2.2.30.\n",
      " |      \n",
      " |      This controls the addition of the -qcov_hsp_perc parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  query\n",
      " |      The sequence to search with.\n",
      " |      \n",
      " |      This controls the addition of the -query parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  query_loc\n",
      " |      Location on the query sequence (Format: start-stop).\n",
      " |      \n",
      " |      This controls the addition of the -query_loc parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  remote\n",
      " |      Execute search remotely?\n",
      " |      \n",
      " |      Incompatible with: gilist, negative_gilist, subject_loc, num_threads, ...\n",
      " |      \n",
      " |      This property controls the addition of the -remote switch, treat this property as a boolean.\n",
      " |  \n",
      " |  reward\n",
      " |      Reward for a nucleotide match (integer, at least zero).\n",
      " |      \n",
      " |      This controls the addition of the -reward parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  searchsp\n",
      " |      Effective length of the search space (integer).\n",
      " |      \n",
      " |      This controls the addition of the -searchsp parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  seqidlist\n",
      " |      Restrict search of database to list of SeqID's.\n",
      " |      \n",
      " |      Incompatible with: gilist, negative_gilist, remote, subject, subject_loc\n",
      " |      \n",
      " |      This controls the addition of the -seqidlist parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  show_gis\n",
      " |      Show NCBI GIs in deflines?\n",
      " |      \n",
      " |      This property controls the addition of the -show_gis switch, treat this property as a boolean.\n",
      " |  \n",
      " |  soft_masking\n",
      " |      Apply filtering locations as soft masks (Boolean, Default = true).\n",
      " |      \n",
      " |      This controls the addition of the -soft_masking parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  strand\n",
      " |      Query strand(s) to search against database/subject.\n",
      " |      \n",
      " |      Values allowed are \"both\" (default), \"minus\", \"plus\".\n",
      " |      \n",
      " |      This controls the addition of the -strand parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  subject\n",
      " |      Subject sequence(s) to search.\n",
      " |      \n",
      " |      Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask\n",
      " |      \n",
      " |      See also subject_loc.\n",
      " |      \n",
      " |      This controls the addition of the -subject parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  subject_loc\n",
      " |      Location on the subject sequence (Format: start-stop).\n",
      " |      \n",
      " |      Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask, remote.\n",
      " |      \n",
      " |      See also subject.\n",
      " |      \n",
      " |      This controls the addition of the -subject_loc parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  sum_statistics\n",
      " |      Use sum statistics.\n",
      " |      \n",
      " |      This property controls the addition of the -sum_statistics switch, treat this property as a boolean.\n",
      " |  \n",
      " |  sum_stats\n",
      " |      Use sum statistics (boolean).\n",
      " |      \n",
      " |      Added in BLAST+ 2.2.30.\n",
      " |      \n",
      " |      This controls the addition of the -sum_stats parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  task\n",
      " |      Task to execute (string, default 'megablast')\n",
      " |      \n",
      " |      Allowed values 'blastn', 'blastn-short', 'dc-megablast', 'megablast' (the default), or 'vecscreen'.\n",
      " |      \n",
      " |      This controls the addition of the -task parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  template_length\n",
      " |      Discontiguous MegaBLAST template length (integer).\n",
      " |      \n",
      " |      Allowed values: 16, 18, 21.\n",
      " |      \n",
      " |      Requires: template_type.\n",
      " |      \n",
      " |      This controls the addition of the -template_length parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  template_type\n",
      " |      Discontiguous MegaBLAST template type (string).\n",
      " |      \n",
      " |      Allowed values: 'coding', 'coding_and_optimal' or 'optimal'.\n",
      " |      Requires: template_length.\n",
      " |      \n",
      " |      This controls the addition of the -template_type parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  ungapped\n",
      " |      Perform ungapped alignment only?\n",
      " |      \n",
      " |      This property controls the addition of the -ungapped switch, treat this property as a boolean.\n",
      " |  \n",
      " |  use_index\n",
      " |      Use MegaBLAST database index (Boolean, Default = False)\n",
      " |      \n",
      " |      This controls the addition of the -use_index parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  version\n",
      " |      Print version number;  ignore other arguments.\n",
      " |      \n",
      " |      This property controls the addition of the -version switch, treat this property as a boolean.\n",
      " |  \n",
      " |  window_masker_db\n",
      " |      Enable WindowMasker filtering using this repeats database (string).\n",
      " |      \n",
      " |      This controls the addition of the -window_masker_db parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  window_masker_taxid\n",
      " |      Enable WindowMasker filtering using a Taxonomic ID (integer).\n",
      " |      \n",
      " |      This controls the addition of the -window_masker_taxid parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  window_size\n",
      " |      Multiple hits window size, use 0 to specify 1-hit algorithm (integer).\n",
      " |      \n",
      " |      This controls the addition of the -window_size parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  word_size\n",
      " |      Word size for wordfinder algorithm.\n",
      " |      \n",
      " |      Integer. Minimum 2.\n",
      " |      \n",
      " |      This controls the addition of the -word_size parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  xdrop_gap\n",
      " |      X-dropoff value (in bits) for preliminary gapped extensions (float).\n",
      " |      \n",
      " |      This controls the addition of the -xdrop_gap parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  xdrop_gap_final\n",
      " |      X-dropoff value (in bits) for final gapped alignment (float).\n",
      " |      \n",
      " |      This controls the addition of the -xdrop_gap_final parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  xdrop_ungap\n",
      " |      X-dropoff value (in bits) for ungapped extensions (float).\n",
      " |      \n",
      " |      This controls the addition of the -xdrop_ungap parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __call__(self, stdin=None, stdout=True, stderr=True, cwd=None, env=None)\n",
      " |      Execute command, wait for it to finish, return (stdout, stderr).\n",
      " |      \n",
      " |      Runs the command line tool and waits for it to finish. If it returns\n",
      " |      a non-zero error level, an exception is raised. Otherwise two strings\n",
      " |      are returned containing stdout and stderr.\n",
      " |      \n",
      " |      The optional stdin argument should be a string of data which will be\n",
      " |      passed to the tool as standard input.\n",
      " |      \n",
      " |      The optional stdout and stderr argument may be filenames (string),\n",
      " |      but otherwise are treated as a booleans, and control if the output\n",
      " |      should be captured as strings (True, default), or ignored by sending\n",
      " |      it to /dev/null to avoid wasting memory (False). If sent to a file\n",
      " |      or ignored, then empty string(s) are returned.\n",
      " |      \n",
      " |      The optional cwd argument is a string giving the working directory\n",
      " |      to run the command from. See Python's subprocess module documentation\n",
      " |      for more details.\n",
      " |      \n",
      " |      The optional env argument is a dictionary setting the environment\n",
      " |      variables to be used in the new process. By default the current\n",
      " |      process' environment variables are used. See Python's subprocess\n",
      " |      module documentation for more details.\n",
      " |      \n",
      " |      Default example usage::\n",
      " |      \n",
      " |          from Bio.Emboss.Applications import WaterCommandline\n",
      " |          water_cmd = WaterCommandline(gapopen=10, gapextend=0.5,\n",
      " |                                       stdout=True, auto=True,\n",
      " |                                       asequence=\"a.fasta\", bsequence=\"b.fasta\")\n",
      " |          print(\"About to run: %s\" % water_cmd)\n",
      " |          std_output, err_output = water_cmd()\n",
      " |      \n",
      " |      This functionality is similar to subprocess.check_output(). In general\n",
      " |      if you require more control over running the command, use subprocess\n",
      " |      directly.\n",
      " |      \n",
      " |      When the program called returns a non-zero error level, a custom\n",
      " |      ApplicationError exception is raised. This includes any stdout and\n",
      " |      stderr strings captured as attributes of the exception object, since\n",
      " |      they may be useful for diagnosing what went wrong.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a representation of the command line object for debugging.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> cline\n",
      " |      WaterCommandline(cmd='water', outfile='temp_water.txt', asequence='asis:ACCCGGGCGCGGT', bsequence='asis:ACCCGAGCGCGGT', gapopen=10, gapextend=0.5)\n",
      " |  \n",
      " |  __setattr__(self, name, value)\n",
      " |      Set attribute name to value (PRIVATE).\n",
      " |      \n",
      " |      This code implements a workaround for a user interface issue.\n",
      " |      Without this __setattr__ attribute-based assignment of parameters\n",
      " |      will silently accept invalid parameters, leading to known instances\n",
      " |      of the user assuming that parameters for the application are set,\n",
      " |      when they are not.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5, stdout=True)\n",
      " |      >>> cline.asequence = \"a.fasta\"\n",
      " |      >>> cline.bsequence = \"b.fasta\"\n",
      " |      >>> cline.csequence = \"c.fasta\"\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      ValueError: Option name csequence was not found.\n",
      " |      >>> print(cline)\n",
      " |      water -stdout -asequence=a.fasta -bsequence=b.fasta -gapopen=10 -gapextend=0.5\n",
      " |      \n",
      " |      This workaround uses a whitelist of object attributes, and sets the\n",
      " |      object attribute list as normal, for these.  Other attributes are\n",
      " |      assumed to be parameters, and passed to the self.set_parameter method\n",
      " |      for validation and assignment.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Make the commandline string with the currently set options.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> str(cline)\n",
      " |      'water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5'\n",
      " |  \n",
      " |  set_parameter(self, name, value=None)\n",
      " |      Set a commandline option for a program (OBSOLETE).\n",
      " |      \n",
      " |      Every parameter is available via a property and as a named\n",
      " |      keyword when creating the instance. Using either of these is\n",
      " |      preferred to this legacy set_parameter method which is now\n",
      " |      OBSOLETE, and likely to be DEPRECATED and later REMOVED in\n",
      " |      future releases.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  parameters = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(NcbiblastnCommandline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf366559",
   "metadata": {},
   "outputs": [],
   "source": [
    "blastn -task blastn -query example/unknown.fa -db db/bacteria/bacteria_nucl -out example/blastn_bacteria.out"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21bbae1a",
   "metadata": {},
   "source": [
    "# Running and analyzing a local BLASTN search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a3cdca8d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cmd_blastn = NcbiblastnCommandline(query='July142020-ALL-R1.fasta',\n",
    "                      out='output_blastn.tab',\n",
    "                      outfmt=6,\n",
    "                      db='db_ref/ref_db_16S_GreenAlgae',\n",
    "                      num_threads = 64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "dc3e50c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastn -out output_blastn.tab -outfmt 6 -query July142020-ALL-R1.fasta -db db_ref/ref_db_16S_GreenAlgae -num_threads 64\n"
     ]
    }
   ],
   "source": [
    "# Get a working command-line\n",
    "print(cmd_blastn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ba01cdb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 625 ms, sys: 436 ms, total: 1.06 s\n",
      "Wall time: 3h 22min 16s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "('', '')"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "cmd_blastn()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57792b7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run BLASTX, and catch STDOUT/STDERR\n",
    "# !! Do not execute cell if skipping computation !!\n",
    "stdout, stderr = cmd_blastn()\n",
    "\n",
    "# Check STDOUT, STDERR\n",
    "print(\"STDOUT: %s\" % stdout)\n",
    "print(\"STDERR: %s\" % stderr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "46583dc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !! If you are skipping computational steps, uncomment the line below !!\n",
    "#blastout = os.path.join('prepped', 'kitasatospora', 'AMK19_00175_blastx_kitasatospora.tab')  # BLAST output\n",
    "\n",
    "# Read BLASTX output\n",
    "results = pd.read_csv(\"output_blastn.tab\", sep=\"\\t\", header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9340fd87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6289929, 12)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0b0434bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A00842:193:HMWFWDRXY:1:2101:18322:1094</td>\n",
       "      <td>NC_009630.1</td>\n",
       "      <td>83.146</td>\n",
       "      <td>89</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>120</td>\n",
       "      <td>1992</td>\n",
       "      <td>1904</td>\n",
       "      <td>1.170000e-15</td>\n",
       "      <td>82.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A00842:193:HMWFWDRXY:1:2101:18322:1094</td>\n",
       "      <td>EF463011.1</td>\n",
       "      <td>83.146</td>\n",
       "      <td>89</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>120</td>\n",
       "      <td>1992</td>\n",
       "      <td>1904</td>\n",
       "      <td>1.170000e-15</td>\n",
       "      <td>82.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A00842:193:HMWFWDRXY:1:2101:18322:1094</td>\n",
       "      <td>NC_038217.1</td>\n",
       "      <td>95.238</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>68</td>\n",
       "      <td>187950</td>\n",
       "      <td>187909</td>\n",
       "      <td>3.270000e-11</td>\n",
       "      <td>67.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A00842:193:HMWFWDRXY:1:2101:18322:1094</td>\n",
       "      <td>MN201586.1</td>\n",
       "      <td>95.238</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>68</td>\n",
       "      <td>155398</td>\n",
       "      <td>155439</td>\n",
       "      <td>3.270000e-11</td>\n",
       "      <td>67.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A00842:193:HMWFWDRXY:1:2101:18322:1094</td>\n",
       "      <td>KX761577.1</td>\n",
       "      <td>95.238</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>68</td>\n",
       "      <td>187950</td>\n",
       "      <td>187909</td>\n",
       "      <td>3.270000e-11</td>\n",
       "      <td>67.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       0            1       2   3   4   5   \\\n",
       "0  A00842:193:HMWFWDRXY:1:2101:18322:1094  NC_009630.1  83.146  89  15   0   \n",
       "1  A00842:193:HMWFWDRXY:1:2101:18322:1094   EF463011.1  83.146  89  15   0   \n",
       "2  A00842:193:HMWFWDRXY:1:2101:18322:1094  NC_038217.1  95.238  42   2   0   \n",
       "3  A00842:193:HMWFWDRXY:1:2101:18322:1094   MN201586.1  95.238  42   2   0   \n",
       "4  A00842:193:HMWFWDRXY:1:2101:18322:1094   KX761577.1  95.238  42   2   0   \n",
       "\n",
       "   6    7       8       9             10    11  \n",
       "0  32  120    1992    1904  1.170000e-15  82.4  \n",
       "1  32  120    1992    1904  1.170000e-15  82.4  \n",
       "2  27   68  187950  187909  3.270000e-11  67.6  \n",
       "3  27   68  155398  155439  3.270000e-11  67.6  \n",
       "4  27   68  187950  187909  3.270000e-11  67.6  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Inspect results table\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "22cf3997",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>subject</th>\n",
       "      <th>pc_identity</th>\n",
       "      <th>aln_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gaps_opened</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A00842:193:HMWFWDRXY:1:2101:18322:1094</td>\n",
       "      <td>NC_009630.1</td>\n",
       "      <td>83.146</td>\n",
       "      <td>89</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>120</td>\n",
       "      <td>1992</td>\n",
       "      <td>1904</td>\n",
       "      <td>1.170000e-15</td>\n",
       "      <td>82.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A00842:193:HMWFWDRXY:1:2101:18322:1094</td>\n",
       "      <td>EF463011.1</td>\n",
       "      <td>83.146</td>\n",
       "      <td>89</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>120</td>\n",
       "      <td>1992</td>\n",
       "      <td>1904</td>\n",
       "      <td>1.170000e-15</td>\n",
       "      <td>82.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A00842:193:HMWFWDRXY:1:2101:18322:1094</td>\n",
       "      <td>NC_038217.1</td>\n",
       "      <td>95.238</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>68</td>\n",
       "      <td>187950</td>\n",
       "      <td>187909</td>\n",
       "      <td>3.270000e-11</td>\n",
       "      <td>67.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A00842:193:HMWFWDRXY:1:2101:18322:1094</td>\n",
       "      <td>MN201586.1</td>\n",
       "      <td>95.238</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>68</td>\n",
       "      <td>155398</td>\n",
       "      <td>155439</td>\n",
       "      <td>3.270000e-11</td>\n",
       "      <td>67.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A00842:193:HMWFWDRXY:1:2101:18322:1094</td>\n",
       "      <td>KX761577.1</td>\n",
       "      <td>95.238</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>68</td>\n",
       "      <td>187950</td>\n",
       "      <td>187909</td>\n",
       "      <td>3.270000e-11</td>\n",
       "      <td>67.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    query      subject  pc_identity  \\\n",
       "0  A00842:193:HMWFWDRXY:1:2101:18322:1094  NC_009630.1       83.146   \n",
       "1  A00842:193:HMWFWDRXY:1:2101:18322:1094   EF463011.1       83.146   \n",
       "2  A00842:193:HMWFWDRXY:1:2101:18322:1094  NC_038217.1       95.238   \n",
       "3  A00842:193:HMWFWDRXY:1:2101:18322:1094   MN201586.1       95.238   \n",
       "4  A00842:193:HMWFWDRXY:1:2101:18322:1094   KX761577.1       95.238   \n",
       "\n",
       "   aln_length  mismatches  gaps_opened  query_start  query_end  subject_start  \\\n",
       "0          89          15            0           32        120           1992   \n",
       "1          89          15            0           32        120           1992   \n",
       "2          42           2            0           27         68         187950   \n",
       "3          42           2            0           27         68         155398   \n",
       "4          42           2            0           27         68         187950   \n",
       "\n",
       "   subject_end       e_value  bitscore  \n",
       "0         1904  1.170000e-15      82.4  \n",
       "1         1904  1.170000e-15      82.4  \n",
       "2       187909  3.270000e-11      67.6  \n",
       "3       155439  3.270000e-11      67.6  \n",
       "4       187909  3.270000e-11      67.6  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define column headers\n",
    "headers = ['query', 'subject',\n",
    "           'pc_identity', 'aln_length', 'mismatches', 'gaps_opened',\n",
    "           'query_start', 'query_end', 'subject_start', 'subject_end',\n",
    "           'e_value', 'bitscore']\n",
    "\n",
    "# Assign headers\n",
    "results.columns = headers\n",
    "\n",
    "# Inspect modified table\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "591ab7b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pc_identity</th>\n",
       "      <th>aln_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gaps_opened</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>6.289929e+06</td>\n",
       "      <td>6.289929e+06</td>\n",
       "      <td>6.289929e+06</td>\n",
       "      <td>6.289929e+06</td>\n",
       "      <td>6.289929e+06</td>\n",
       "      <td>6.289929e+06</td>\n",
       "      <td>6.289929e+06</td>\n",
       "      <td>6.289929e+06</td>\n",
       "      <td>6.289929e+06</td>\n",
       "      <td>6.289929e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>9.049531e+01</td>\n",
       "      <td>1.061238e+02</td>\n",
       "      <td>9.981368e+00</td>\n",
       "      <td>1.038350e+00</td>\n",
       "      <td>1.592050e+01</td>\n",
       "      <td>1.204425e+02</td>\n",
       "      <td>3.392174e+04</td>\n",
       "      <td>3.392104e+04</td>\n",
       "      <td>6.004135e-09</td>\n",
       "      <td>1.330045e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.149892e+00</td>\n",
       "      <td>3.682165e+01</td>\n",
       "      <td>6.816648e+00</td>\n",
       "      <td>1.803333e+00</td>\n",
       "      <td>2.431531e+01</td>\n",
       "      <td>3.415107e+01</td>\n",
       "      <td>9.292986e+04</td>\n",
       "      <td>9.292980e+04</td>\n",
       "      <td>1.133176e-06</td>\n",
       "      <td>4.212917e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>7.337700e+01</td>\n",
       "      <td>2.500000e+01</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.800000e+01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3.860000e-75</td>\n",
       "      <td>4.170000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>8.686100e+01</td>\n",
       "      <td>7.500000e+01</td>\n",
       "      <td>4.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.700000e+01</td>\n",
       "      <td>4.840000e+02</td>\n",
       "      <td>4.840000e+02</td>\n",
       "      <td>1.460000e-39</td>\n",
       "      <td>1.000000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>9.059800e+01</td>\n",
       "      <td>1.090000e+02</td>\n",
       "      <td>9.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.000000e+00</td>\n",
       "      <td>1.350000e+02</td>\n",
       "      <td>1.116000e+03</td>\n",
       "      <td>1.107000e+03</td>\n",
       "      <td>4.120000e-30</td>\n",
       "      <td>1.300000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>9.440600e+01</td>\n",
       "      <td>1.420000e+02</td>\n",
       "      <td>1.500000e+01</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>2.200000e+01</td>\n",
       "      <td>1.510000e+02</td>\n",
       "      <td>4.466900e+04</td>\n",
       "      <td>4.466900e+04</td>\n",
       "      <td>3.230000e-21</td>\n",
       "      <td>1.610000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.000000e+02</td>\n",
       "      <td>1.770000e+02</td>\n",
       "      <td>3.900000e+01</td>\n",
       "      <td>2.000000e+01</td>\n",
       "      <td>1.240000e+02</td>\n",
       "      <td>1.510000e+02</td>\n",
       "      <td>4.263916e+06</td>\n",
       "      <td>4.263944e+06</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>2.790000e+02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        pc_identity    aln_length    mismatches   gaps_opened   query_start  \\\n",
       "count  6.289929e+06  6.289929e+06  6.289929e+06  6.289929e+06  6.289929e+06   \n",
       "mean   9.049531e+01  1.061238e+02  9.981368e+00  1.038350e+00  1.592050e+01   \n",
       "std    5.149892e+00  3.682165e+01  6.816648e+00  1.803333e+00  2.431531e+01   \n",
       "min    7.337700e+01  2.500000e+01  0.000000e+00  0.000000e+00  1.000000e+00   \n",
       "25%    8.686100e+01  7.500000e+01  4.000000e+00  0.000000e+00  1.000000e+00   \n",
       "50%    9.059800e+01  1.090000e+02  9.000000e+00  0.000000e+00  3.000000e+00   \n",
       "75%    9.440600e+01  1.420000e+02  1.500000e+01  2.000000e+00  2.200000e+01   \n",
       "max    1.000000e+02  1.770000e+02  3.900000e+01  2.000000e+01  1.240000e+02   \n",
       "\n",
       "          query_end  subject_start   subject_end       e_value      bitscore  \n",
       "count  6.289929e+06   6.289929e+06  6.289929e+06  6.289929e+06  6.289929e+06  \n",
       "mean   1.204425e+02   3.392174e+04  3.392104e+04  6.004135e-09  1.330045e+02  \n",
       "std    3.415107e+01   9.292986e+04  9.292980e+04  1.133176e-06  4.212917e+01  \n",
       "min    2.800000e+01   1.000000e+00  1.000000e+00  3.860000e-75  4.170000e+01  \n",
       "25%    9.700000e+01   4.840000e+02  4.840000e+02  1.460000e-39  1.000000e+02  \n",
       "50%    1.350000e+02   1.116000e+03  1.107000e+03  4.120000e-30  1.300000e+02  \n",
       "75%    1.510000e+02   4.466900e+04  4.466900e+04  3.230000e-21  1.610000e+02  \n",
       "max    1.510000e+02   4.263916e+06  4.263944e+06  2.000000e-03  2.790000e+02  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show a summary of the results table data\n",
    "results.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3a4c9a0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0          NC_009630.1\n",
      "1           EF463011.1\n",
      "2          NC_038217.1\n",
      "3           MN201586.1\n",
      "4           KX761577.1\n",
      "              ...     \n",
      "6289924     KX602105.1\n",
      "6289925     KX602101.1\n",
      "6289926     KX602099.1\n",
      "6289927     KX602074.1\n",
      "6289928     KX602081.1\n",
      "Name: subject, Length: 6289929, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# Show all subject matches\n",
    "print(results['subject'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63ba1742",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new column describing how long the alignment is on the query sequence\n",
    "qaln_length = abs(results['query_end'] - results['query_start']) + 1\n",
    "print(qaln_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60360b7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add qaln_length to the results table as a new column\n",
    "results['qaln_length'] = qaln_length\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ec06493",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a scatterplot\n",
    "results.plot.scatter('pc_identity', 'e_value')\n",
    "plt.title(\"E value vs %identity\");              # add a title to the plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fa45e58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SOLUTION - EXERCISE 01\n",
    "# !! Do not execute this cell if skipping computational step !!\n",
    "\n",
    "# We can reuse the directories and db, but need to define new input/output filenames\n",
    "query = os.path.join(datadir, 'lantibiotic.fasta')                                   # query sequence(s)\n",
    "blastout = os.path.join(outdir, 'lantibiotic_blastx_kitasatospora.tab')              # BLAST output\n",
    "\n",
    "# Create command-line for BLASTX\n",
    "cmd_blastx = NcbiblastxCommandline(query=query, out=blastout, outfmt=6, db=db)\n",
    "\n",
    "# Run BLASTX, and catch STDOUT/STDERR\n",
    "stdout, stderr = cmd_blastx()\n",
    "\n",
    "# Check STDOUT, STDERR\n",
    "print(\"STDOUT: %s\" % stdout)\n",
    "print(\"STDERR: %s\" % stderr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8acd462b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !! Uncomment the line below, if skipping computational step !!\n",
    "# blastout = os.path.join('prepped', 'kitasatospora', 'lantibiotic_blastx_kitasatospora.tab')\n",
    "\n",
    "# Read BLASTX output, and reuse the column headers defined earlier\n",
    "results = pd.read_csv(blastout, sep=\"\\t\", header=None)\n",
    "results.columns = headers\n",
    "\n",
    "# Create a scatterplot\n",
    "results.plot.scatter('bitscore', 'pc_identity')\n",
    "plt.title(\"%identity vs bitscore\");  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
